css/css-view-transitions/fractional-* fails due to not pixel snapping view transition renderers.

mattwoodrow opened the pull request:
#### 72ba48fd55e6120b28f47394c5ea63373cdd53ed
<pre>
css/css-view-transitions/fractional-* fails due to not pixel snapping view transition renderers.
<a href="https://bugs.webkit.org/show_bug.cgi?id=293273">https://bugs.webkit.org/show_bug.cgi?id=293273</a>
&lt;<a href="https://rdar.apple.com/151661047">rdar://151661047</a>&gt;

Reviewed by Simon Fraser.

Moves the duplicated state in CapturedElement into a separate struct, and use
that to simplify parameters of copyElementBaseProperties.

Compute and remove the fractional component of the view transition transform,
and pass it through to the snapshot rendering.

Snap the layer rects to pixels, so that the fractional component gets computed
and passed down the inner layer, as it does for non-VT rendering.

* LayoutTests/TestExpectations:
* LayoutTests/platform/ios/TestExpectations:
* Source/WebCore/dom/ViewTransition.cpp:
(WebCore::captureOverflowRect):
(WebCore::snapshotElementVisualOverflowClippedToViewport):
(WebCore::ViewTransition::captureOldState):
(WebCore::ViewTransition::updatePropertiesForGroupPseudo):
(WebCore::ViewTransition::captureNewState):
(WebCore::ViewTransition::setupDynamicStyleSheet):
(WebCore::snapTransformationTranslationToDevicePixels):
(WebCore::ViewTransition::copyElementBaseProperties):
(WebCore::ViewTransition::updatePseudoElementStylesRead):
(WebCore::ViewTransition::updatePseudoElementRenderers):
* Source/WebCore/dom/ViewTransition.h:
* Source/WebCore/platform/graphics/transforms/TransformationMatrix.h:
(WebCore::TransformationMatrix::hasPerspective const):
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::outputPaintOrderTreeRecursive):
(WebCore::RenderLayer::filtersForPainting const): Deleted.
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::updateGeometry):
* Source/WebCore/rendering/updating/RenderTreeUpdaterViewTransition.cpp:
(WebCore::createRendererIfNeeded):

Canonical link: <a href="https://commits.webkit.org/295401@main">https://commits.webkit.org/295401@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d3a4cd872d3554787aeead29b992040147536912

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/104969 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/24683 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/15104 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/110185 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/55648 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/107010 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/25089 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/33227 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/79704 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/107975 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/19519 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/94735 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/60011 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/19263 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/12813 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/55026 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/88982 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/12859 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/112672 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/32134 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/23638 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/88785 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/32499 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/90961 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/88413 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22538 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/33306 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/11088 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/27466 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/32058 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/37429 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/31851 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/35192 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/33410 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/dom/ViewTransition.cpp:
```diff
@@ -405,7 +405,7 @@ static LayoutRect captureOverflowRect(RenderLayerModelObject& renderer)
     }
 
-    return renderer.layer()->calculateLayerBounds(renderer.layer(), LayoutSize(), { RenderLayer::IncludeFilterOutsets, RenderLayer::ExcludeHiddenDescendants, RenderLayer::IncludeCompositedDescendants, RenderLayer::PreserveAncestorFlags });
+    return LayoutRect(encloseRectToDevicePixels(renderer.layer()->calculateLayerBounds(renderer.layer(), LayoutSize(), { RenderLayer::IncludeFilterOutsets, RenderLayer::ExcludeHiddenDescendants, RenderLayer::IncludeCompositedDescendants, RenderLayer::PreserveAncestorFlags }), renderer.document().deviceScaleFactor()));
```
I find it easier to debug lines like this when the result of the inner function is stored in a local variable.

smfr commented on Source/WebCore/dom/ViewTransition.h:
```diff
@@ -68,19 +68,22 @@ enum class ViewTransitionPhase : uint8_t {
 struct CapturedElement {
     WTF_MAKE_TZONE_ALLOCATED(CapturedElement);
 public:
+    struct ResolvedStyle {
```
Not just "Style"

smfr commented on Source/WebCore/platform/graphics/transforms/TransformationMatrix.h:
```diff
@@ -315,6 +316,10 @@ class TransformationMatrix {
 
+    // Rounds the x/y translation components to the nearest pixel (for non-perspective)
+    // transforms, and returns the subpixel offset that was removed.
+    LayoutSize round2dTranslationComponents(float deviceScaleFactor);
```
I don't think LayoutUnits should ever show up in this header; this rounding should move elsewhere.

nt1m commented on Source/WebCore/dom/ViewTransition.h:
```diff
@@ -68,19 +68,22 @@ enum class ViewTransitionPhase : uint8_t {
 struct CapturedElement {
     WTF_MAKE_TZONE_ALLOCATED(CapturedElement);
 public:
+    struct Style {
```
I think Simon meant that these are not just "style" related things.

Not really sure of a better name here though

mattwoodrow replied:
That is a much better interpretation of the comment.

How about `ElementState`? The spec steps that fill these in are 'capture the new/old state' (for a captured element).

nt1m replied:
`CaptureState` is a good option too based on the prose in the spec:

> Set capture’s [old image](https://drafts.csswg.org/css-view-transitions-1/#captured-element-old-image) to the result of [capturing the image](https://drafts.csswg.org/css-view-transitions-1/#capture-the-image) of element.

mattwoodrow replied:
Alright, gone with that!

nt1m commented on Source/WebCore/dom/ViewTransition.cpp:
```diff
@@ -825,7 +822,21 @@ void ViewTransition::clearViewTransition()
+}
+
+
+void ViewTransition::copyElementBaseProperties(RenderLayerModelObject& renderer, CapturedElement::CaptureState& output)
```
```suggestion
void ViewTransition::copyElementBaseProperties(RenderLayerModelObject& renderer, CapturedElement::State& output)
```

This would read slightly nicer now that I see the usages. I'd just rename to `State` as `CapturedElement` provides enough context

mattwoodrow committed:
css/css-view-transitions/fractional-* fails due to not pixel snapping view transition renderers.
https://bugs.webkit.org/show_bug.cgi?id=293273
<rdar://151661047>

Reviewed by Simon Fraser.

Moves the duplicated state in CapturedElement into a separate struct, and use
that to simplify parameters of copyElementBaseProperties.

Compute and remove the fractional component of the view transition transform,
and pass it through to the snapshot rendering.

Snap the layer rects to pixels, so that the fractional component gets computed
and passed down the inner layer, as it does for non-VT rendering.

* LayoutTests/TestExpectations:
* LayoutTests/platform/ios/TestExpectations:
* Source/WebCore/dom/ViewTransition.cpp:
(WebCore::captureOverflowRect):
(WebCore::snapshotElementVisualOverflowClippedToViewport):
(WebCore::ViewTransition::captureOldState):
(WebCore::ViewTransition::updatePropertiesForGroupPseudo):
(WebCore::ViewTransition::captureNewState):
(WebCore::ViewTransition::setupDynamicStyleSheet):
(WebCore::snapTransformationTranslationToDevicePixels):
(WebCore::ViewTransition::copyElementBaseProperties):
(WebCore::ViewTransition::updatePseudoElementStylesRead):
(WebCore::ViewTransition::updatePseudoElementRenderers):
* Source/WebCore/dom/ViewTransition.h:
* Source/WebCore/platform/graphics/transforms/TransformationMatrix.h:
(WebCore::TransformationMatrix::hasPerspective const):
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::outputPaintOrderTreeRecursive):
(WebCore::RenderLayer::filtersForPainting const): Deleted.
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::updateGeometry):
* Source/WebCore/rendering/updating/RenderTreeUpdaterViewTransition.cpp:
(WebCore::createRendererIfNeeded):

Canonical link: https://commits.webkit.org/295401@main
